Update llama data preparation

nsakkine opened the pull request:
Updates llama example data preparation by

- Cleaning-up script
- Enforcing use of HuggingFace tokenizers to match training scripts
- Enabling 'fineweb' dataset

and improves training scripts' consistency (also adds support for checkpointing) as well as README quality.

nsakkine committed:
Tentative update to data preparing process: cleaned-up script, enforced HF tokenizer choice and added 'fineweb' dataset.

lcskrishna commented on examples/llama/bookcorpus/download.py:
```diff
@@ -11,4 +11,4 @@
     dataset = load_dataset("bookcorpus", split="train", trust_remote_code=True)
-    dataset.to_json(out_dir / "bookcorpus_megatron.json")
\ No newline at end of file
+    dataset.to_json(out_dir / "data.jsonl")
```
any difference of size between .json and .jsonl, please don't change the default names as there are other automations that might get affected.

nsakkine replied:
Just name changes (jsonl refers explicitly to json-lines format). With the change, we can use the same preprocessing steps, and thus avoid duplicate code in `prepare_data.sh`. Please point me to the automations that you refer to, I only see that CI passes.

lcskrishna replied:
okay please reach out offline.

lcskrishna commented on examples/llama/prepare_dataset.sh:
```diff
@@ -1,43 +1,51 @@
+echo "Preprocessing '${DATASET}' dataset using '${TOKENIZER_MODEL}' tokenizer..."
+python3 tools/preprocess_data.py \
+    --input "${DATASET_PATH}"  \
+    --tokenizer-type "HuggingFaceTokenizer" \
```
Tokenizer should be updated based on the dataset here. bookcorpus is using GPTSentensePieceTokenizer and so on.

nsakkine replied:
Can you pinpoint me to the reason GPTSentencePieceTokenizer in particular needs to be used with bookcorpus?

lcskrishna replied:
No specific reason. The previous ones are using GPTSentencePieceTokenizer which you have already deleted. Is there a specific reason to switch to HuggingfaceTokenizer rather than LlamaTokenizer present in the repo already? What benefit does it give. You might have to test it every possible cases if we switch it.

nsakkine replied:
HuggingFace tokenizer is the default for both training scripts in `rocm_dev` branch. This change only improves consistency in the tokenizer choice.

lcskrishna commented on examples/llama/prepare_dataset.sh:
```diff
@@ -1,43 +1,51 @@
-if ! [ -f "$TOKENIZER_MODEL" ]; then
-wget -O $TOKENIZER_MODEL https://huggingface.co/NousResearch/Llama-2-7b-chat-hf/resolve/main/tokenizer.model
+# Check input
+if [ -z ${TOKENIZER_MODEL+x} ]; then
```
I would suggest NousResearch tokenizer model should be kept default. Otherwise please update the README along with the links for the tokenizer model.

nsakkine replied:
Sure, we can have a default tokenizer, will change

nsakkine replied:
Decided after all to modify README and not implement a default tokenizer. Can revert if Llama2 training can be considered as a default, otherwise I do not see the point of having a default tokenizer, since it differs between the training scripts.

lcskrishna reviewed (changes requested):
Few comments, we can discuss offline if needed.

nsakkine committed:
Updated README, fixed data preparation problems and cleaned-up training
scripts.

nsakkine commented:
Fixed some issues with dataset preparation. Updated README (also regarding training scripts' usage, but could not test multi-node setup). Some usability changes to training scripts. Tested README instructions. Plan on adding checkpointing next. @lcskrishna please review these changes.

gurpreet-dhami reviewed (commented):
@nsakkine : Why is this section removed ? How will it parse the passed arguments otherwise ?

gurpreet-dhami commented on examples/llama/train_llama2.sh:
```diff
@@ -19,19 +19,6 @@ export RCCL_MSCCL_ENABLE=0
 
-
-# parsing input arguments
-for ARGUMENT in "$@"
```
@nsakkine : Why is this section removed ? How will it parse the passed arguments otherwise ?

nsakkine replied:
What this essentially does is that it reads in positional parameters, parses them and exports them as environment variables. I am proposing to remove this because I have not seen in any example that the script would have used by passing positional parameters to it. Moreover, it seems to me that this serves the same purpose as directly passing information as environment variables which is the way in which the script used in all examples I have seen. Please do indicate if I am mistaken here.

gurpreet-dhami replied:
@nsakkine  : We are using this in DLM already this way. 
Please suggest, how can we test different configurations otherwise ? like BS, TP, MBS, etc..

https://github.com/ROCm/DeepLearningModels/blob/077c263b91d1ab3c42e5e381a3eb55edc56ce158/scripts/pyt_megatron_lm_train/run.sh#L71
https://github.com/ROCm/DeepLearningModels/blob/077c263b91d1ab3c42e5e381a3eb55edc56ce158/models.json#L5022

nsakkine replied:
This change (i.e. removing positional parameters parsing) does not affect what you linked since they use environment variables to pass information to the training scripts. But since there might be some downstream use cases relying on positional parameters, I think you are right in that it is better to restore the parsing logic rather than risk breaking something.

aivanni replied:
Maybe by itself it is ok idea to allow passing the same parameters both as environment variable and positional argument, but doing it this way blocks passing through arbitrary arguments from the megatron's list https://github.com/ROCm/Megatron-LM/blob/rocm_dev/megatron/training/arguments.py . Without this argument parsing block we could have something like (with @ at the end) so any megatron's arguments could passed:

```
run_cmd="
    torchrun $DISTRIBUTED_ARGS pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        $EXTRA_ARGS \
        $TRAIN_ARGS \
        $CKPT_LOAD_ARGS \
        $@
"
```

nsakkine replied:
You are naturally right. But unless there is now a clear vision on what to prefer @gurpreet-dhami @lcskrishna I would perhaps be inclined to suggest that we freeze this PR, get it merged, and despite loosing some momentum, think about what to do with the positional parameters separately.

gurpreet-dhami reviewed (commented):
I would suggest not to mix up the cleanup and data preparation tasks.
It prolongs the PR to get reviewed and merged. You may create a separate PR.

nsakkine commented:
Thank you @gurpreet-dhami for your review. I am at the moment only proposing changes to the training scripts which relate to using real data for training. I believe this would be the right place to do so since otherwise the proposed changes to data preparation would break the workflows described in the original README. Making two separate PRs would then leave `rocm_dev` branch temporarily in an inconsistent state.

nsakkine committed:
Fixed details in README

nsakkine committed:
Restored positional parameters parsing. Added checkpointing.

gurpreet-dhami commented on examples/llama/train_llama2.sh:
@nsakkine 
Do you expect to pass TOKENIZER_MODEL along with TOKENIZER_TYPE. How are you planning to run Llama2Tokenizer in this implementation ? 
Earlier,   TOKENIZER_TYPE = Llama2Tokenizer, was enough.

nsakkine replied:
What you described is one way. Other way would be to use the default HuggingFace NousResearch/Llama-2-7b-chat-hf tokenizer. From the point of view of pretraining, AFAIK there is not much of a difference: some special tokens are defined differently in Llama2Tokenizer. I would assume that as long as one is consistent with the tokenizer choice everything is good. Performance should not be affected. Having said this, I reverted to the earlier implementation which allows Llama2Tokenizer special handling just in case there would be some difference I do not see.

nsakkine committed:
Restored Llama2Tokenizer

gurpreet-dhami commented:
@nsakkine 
DATASET=bookcorpus DATA_DIR=bookcorpus TOKENIZER_MODEL=NousResearch/Llama-2-7b-chat-hf **TOKENIZER_TYPE=GPTSentencePieceTokenizer** bash examples/llama/prepare_dataset.sh 

  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/sentencepiece/__init__.py", line 961, in Load
    return self.LoadFromFile(model_file)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/sentencepiece/__init__.py", line 316, in LoadFromFile
    return _sentencepiece.SentencePieceProcessor_LoadFromFile(self, arg)
OSError: Not found: "NousResearch/Llama-2-7b-chat-hf": No such file or directory Error 

**But it worked with this command.**
DATASET=bookcorpus DATA_DIR=bookcorpus TOKENIZER_MODEL=**~/.cache/huggingface/hub/models--NousResearch--Llama-2-7b-chat-hf/snapshots/351844e75ed0bcbbe3f10671b3c808d2b83894ee/tokenizer.model** TOKENIZER_TYPE=GPTSentencePieceTokenizer bash examples/llama/prepare_dataset.sh

So, if Tokenizer type is not Huggingface, Then I think TOKENIZER_MODEL parameter is a must required parameter. So we can add a check for that in script.

gurpreet-dhami commented on examples/llama/train_llama3.sh:
```diff
@@ -222,20 +199,37 @@ DATA_ARGS="
+    "
+fi
+
+CKPT_LOAD_ARGS="--exit-on-missing-checkpoint \
```
@nsakkine  
If I don't set the LOAD_CKPT arg. Then it gives this error. Please have a look.

  File "/myworkspace/Megatron-LM/megatron/training/initialize.py", line 65, in initialize_megatron
    assert args.load is not None, "--use-checkpoint-args requires --load argument"
AssertionError: --use-checkpoint-args requires --load argument

nsakkine replied:
Should be fixed in an upcoming commit

gurpreet-dhami commented on examples/llama/train_llama3.sh:
```diff
@@ -222,20 +199,37 @@ DATA_ARGS="
 "
-#  --save $CHECKPOINT_PATH \
+if [ ! -z {$LOAD_CKPT_PATH+x} ]; then
+    CKPT_LOAD_ARGS="$CKPT_LOAD_ARGS --load ${LOAD_CKPT_PATH}"
```
@nsakkine 
This condition doesn't work  - 
I am getting this error even if I don't set the LOAD_CKPT_PATH.
- pretrain_gpt.py: error: argument --load: expected one argument

nsakkine replied:
This was a typo, fixed and soon pushing the change

gurpreet-dhami committed:
add tokenizer type as an agrument

gurpreet-dhami commented on examples/llama/train_llama2.sh:
```diff
@@ -232,20 +210,40 @@ DATA_ARGS="
+    --eval-iters -1
+"
+
+if [ -n "$SAVE_CKPT_PATH" ]; then
```
@nsakkine  , I added this change. As it is simpler and essentially does the same thing. 
[ -n "$LOAD_CKPT_PATH" ] → checks if LOAD_CKPT_PATH is set and not empty.

-if [ ! -z ${SAVE_CKPT_PATH+x} ]; then
+if [ -n "$SAVE_CKPT_PATH" ]; then

nsakkine replied:
Thanks for simplifying

gurpreet-dhami commented:
@nsakkine :  Please run the llama2 and llama3 script with tokenizer type Huggingface and LLama2 Tokenizer. and attach the passing logs. It seems like this change is not being tested at all.

nsakkine commented:
> @nsakkine DATASET=bookcorpus DATA_DIR=bookcorpus TOKENIZER_MODEL=NousResearch/Llama-2-7b-chat-hf **TOKENIZER_TYPE=GPTSentencePieceTokenizer** bash examples/llama/prepare_dataset.sh
> 
> File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/sentencepiece/**init**.py", line 961, in Load return self.LoadFromFile(model_file) File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/sentencepiece/**init**.py", line 316, in LoadFromFile return _sentencepiece.SentencePieceProcessor_LoadFromFile(self, arg) OSError: Not found: "NousResearch/Llama-2-7b-chat-hf": No such file or directory Error
> 
> **But it worked with this command.** DATASET=bookcorpus DATA_DIR=bookcorpus TOKENIZER_MODEL=**~/.cache/huggingface/hub/models--NousResearch--Llama-2-7b-chat-hf/snapshots/351844e75ed0bcbbe3f10671b3c808d2b83894ee/tokenizer.model** TOKENIZER_TYPE=GPTSentencePieceTokenizer bash examples/llama/prepare_dataset.sh
> 
> So, if Tokenizer type is not Huggingface, Then I think TOKENIZER_MODEL parameter is a must required parameter. So we can add a check for that in script.

True, if you specify a non-HF tokenizer then you need to pass additional information (not limited only to `TOKENIZER_MODEL`) which creates additional complications. My question would then be that do we need to support all possible tokenizers or rather that we provide an example script which works with selected tokenizers? Presently, the script works with any HF tokenizer and all tokenizers which require only to specify `TOKENIZER_TYPE` and `TOKENIZER_MODEL`  (provided that you have access to the latter). Moreover, `TOKENIZER_MODEL` is in fact required at the moment.

nsakkine committed:
Fixed checkpoint loading

gurpreet-dhami commented on examples/llama/train_llama2.sh:
```diff
@@ -231,21 +206,39 @@ DATA_ARGS="
+    "
+fi
+
+CKPT_LOAD_ARGS=""
```
Why not directly appending it in OUTPUT args, similar to save checkpoint ? 
or else separate out the save args into CKPT_SAVE_ARGS too.

OUTPUT_ARGS="$OUTPUT_ARGS \
        --save-interval $SAVE_INTERVAL \
        --eval-interval $EVAL_INTERVAL \
        --ckpt-format $CKPT_FORMAT \
        --save $SAVE_CKPT_PATH
    "

nsakkine committed:
Changed script defaults to prevent errors due to bad configurations.
Updated README accordingly.

nsakkine commented:
Made changes according to the comments above and changed defaults to enable error free out-of-the-box experience. Checkpointing still requires [this](https://github.com/ROCm/Megatron-LM/pull/64) to be merged. When that is done, can proceed to confirm that also checkpointing works.

lcskrishna reviewed (approved):
LGTM. Let's finish the testing of checkpoint stuff and can merge the PR after that. @gurpreet-dhami Please verify nothing is broken.

vidushi8 commented on examples/llama/readme.md:
```diff
@@ -1,168 +1,130 @@
+    --group-add video --cap-add SYS_PTRACE --security-opt seccomp=unconfined --privileged \
+    -v .:/root/Megatron-LM \
+    --shm-size 64G \
+    rocm/pytorch-training:latest bash
```
We can directly point to rocm/megatron-lm:latest image, which has all the megatron dependencies installed.

nsakkine replied:
Kept `rocm/pytorch-training:latest` as default, because users of the repository do not want a fixed ROCm/Megatron-LM version. But made changes to instruct users in `rocm/megatron-lm:latest` usage as well.

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,168 +1,130 @@
+pip install .
+```
+
+in `/root/Megatron-LM` to install megatron package.
````
Same as above comment, we do not require extra pip install with the megatron-lm docker image.

nsakkine replied:
See above

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,168 +1,130 @@
+
 ```bash
-TEE_OUTPUT=1 MBS=2 BS=64 TP=8 TE_FP8=0 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
+./examples/llama/train_llama2.sh
````
Explicit command line argument in examples, makes it easier for user to understand the test config and knobs which can be toggled. Removing command line parameters is not a good idea IMO.

nsakkine replied:
Made changes to accommodate a richer example with non-default settings.

vidushi8 replied:
Please include `bash` before train scripts.

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,168 +1,130 @@
+
 ```bash
-TEE_OUTPUT=1 MBS=2 BS=16 TP=1 TE_FP8=0 FSDP=1 SEQ_LENGTH=8192 bash examples/llama/train_llama2.sh
+FSDP=1 ./examples/llama/train_llama3.sh
````
Similar comment as above. 
Also, this command will not work as it is. 
Recompute is necessary for 70B models with FSDP. 
Please refer to this PR (https://github.com/ROCm/Megatron-LM/pull/65) which tests FSDP and cleansup scripts/readme with working examples.

nsakkine replied:
See above

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,168 +1,130 @@
-- **Node Rank:**  
-  Set the rank of each node (0 for master, 1 for the first worker node, etc.):
-  ```bash
-  NODE_RANK="${NODE_RANK:-0}"
````
These multi-node variables are important to explicitly list and explain. We should not delete them. This PR (https://github.com/ROCm/Megatron-LM/pull/65) adds multi-node llama support, fixing readme and training scripts.

nsakkine replied:
I agree, restored these

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,168 +1,130 @@
   ```
 
-## 3. Configurations in Script (`Megatron/examples/llama`)
+where `address` is the master node ip address.
````
PR (https://github.com/ROCm/Megatron-LM/pull/65) adds a working example for multi-node llama training. Multi-node requires adding --data-cache-path variable.

nsakkine replied:
Proposing to leave `--data-cache-path` change outside this work, but when doing so, there is a hurry to merge PR#65 to avoid leaving the instructions in an undefined state.

nsakkine committed:
Updated README

vidushi8 commented:
Thanks for addressing the comments @nsakkine. PR LGTM.

aivanni commented on examples/llama/readme.md:
```diff
@@ -1,172 +1,150 @@
 
 ### 2.1 Single Node Training
-To run the training on a single node, go to Megatron-LM folder, use the following command:
+To run training on a single MI300X node, go to ROCm/Megatron-LM repository root and run
```
Are only MI300X intended ? What about MI250X ?

nsakkine replied:
Good point, not sure about "backward compatibility", but for future reference, leaving MI300X out would indeed be better after all.

aivanni commented on examples/llama/readme.md:
````diff
@@ -1,172 +1,150 @@
+on host and update
+
 ```bash
 export NCCL_SOCKET_IFNAME=ens50f0np0
````
shall we explicitly say that `ens50f0np0` here is supposed to be one of the outputs of `ip a`?

nsakkine replied:
Agree to keep as is

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -61,25 +46,50 @@ else
     echo "Single node setup, skipping NCCL and GLOO socket interface settings."
 fi
 
+# parsing input arguments
```
If you moved this block of code down here then the parameters with default values above are not covered anymore by this parsing logic e.g. `EXP_NAME`, `NNODES` etc

nsakkine replied:
True, will fix this

aivanni commented on examples/llama/train_llama3.sh:
```diff
@@ -59,24 +46,40 @@ else
     echo "Single node setup, skipping NCCL and GLOO socket interface settings."
 fi
 
+# parsing input arguments
```
Same here, please pay attention that parameters from above this parsing block of code are not affected anymore

nsakkine replied:
See above

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -231,21 +206,39 @@ DATA_ARGS="
 "
-
-if [ "$MOCK_DATA" -eq 1 ];then
+if [ -z ${DATA_PATH+x} ]; then
```
Why do we have `+x` here? Omitting it should follow mock data path also in case of empty `DATA_PATH` and not just unset

aivanni replied:
or it can be converted to `[ -n "$DATA_PATH" ]` with swapping if and else parts

nsakkine replied:
I think it is better that an invalid `DATA_PATH` (like an empty one) leads to a (hopefully) more verbose error through `pretrain_gpt.py` call.

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -231,21 +206,39 @@ DATA_ARGS="
 
 OUTPUT_ARGS="
     --log-interval 1 \
     --save-interval 5000 \
```
perhaps `--save-interval` is not needed here as it is added inside `if [ -n "$SAVE_CKPT_PATH" ]`

nsakkine replied:
Yes, thanks for spotting this

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -231,21 +206,39 @@ DATA_ARGS="
     --log-throughput \
     --no-save-optim \
-    --eval-iters -1   
+    --eval-iters -1
```
Now always setting `--eval-iters = -1 ` will skip evaluation even if `--eval-interval` is set. Shall we introduce another env var e.g. `EVAL_ITERS` to enable setting `--eval-iters`? and `--log-interval` is something users may want to change too.

gurpreet-dhami replied:
@nsakkine ? what do you think ? Do you to incorporate this one ?

nsakkine replied:
I have incorporated this change

gurpreet-dhami commented:
@nsakkine  : This PR seems to have merge conflicts with rocm_dev. Can you please resolve those?

aivanni commented on examples/llama/readme.md:
```diff
@@ -185,15 +163,10 @@ If you're running multi-node training, update the following environment variable
 
 - **TOTAL_ITERS:**  
-  Set the total number of iterations (default: 10).
+  Sets the total number of iterations (default: 10).
```
default for llama2 is `5` atm

nsakkine replied:
Fixing this by removing the statement about the default

aivanni commented on examples/llama/train_llama3.sh:
```diff
@@ -59,24 +46,40 @@ else
-FSDP="${FSDP:-1}"
+FSDP="${FSDP:-0}"
 RECOMPUTE="${RECOMPUTE:-0}"
+TOKENIZER_TYPE="${TOKENIZER_TYPE:-HuggingFaceTokenizer}"
```
`TOKENIZER_TYPE` defined but not used. Instead `HuggingFaceTokenizer` is hardcoded now in `DATA_ARGS`.

nsakkine replied:
Fixing this

nsakkine committed:
Fixed bugs

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -61,25 +46,50 @@ else
+        exit 1
+    fi
+    TOKENIZER_MODEL="/tmp/tokenizer.model"
+    wget -O "$TOKENIZER_MODEL" https://huggingface.co/NousResearch/Llama-2-7b-chat-hf/resolve/main/tokenizer.model
```
Shall we refer to `Llama-2-7b-hf` model for consistency, as for Llama3 script we refer to base model?

nsakkine replied:
Agreed to keep as is

aivanni commented on examples/llama/prepare_dataset.sh:
```diff
@@ -1,43 +1,46 @@
+# Setup variables
+DATA_DIR="${DATA_DIR:-data}"
+DATASET="${DATASET:-bookcorpus}" # one of wiki,bookcorpus,fineweb
+DATASET_PATH="${DATA_DIR}/data.jsonl"
```
Nit. Maybe worth making `DATA_DIR="$DATA_DIR/$DATASET"` in case one wants to have several datasets ? Really nit, because the user can always specify different `DATA_DIR` corresponding to each dataset

nsakkine replied:
I think `DATA_DIR` offers sufficient flexibility (as you also stated)

nsakkine committed:
Merge branch 'rocm_dev' into update_llama_example_prepare_data_script

nsakkine committed:
Updated scripts with bug fixes and by adding evaluation possibility

aivanni commented on examples/llama/train_llama3.sh:
```diff
@@ -77,10 +77,11 @@ RECOMPUTE="${RECOMPUTE:-0}"
 SAVE_INTERVAL="${SAVE_INTERVAL:-5000}"
+EVAL_ITERS="${EVAL_ITERS:-'-1'}"
 CKPT_FORMAT="${CKPT_FORMAT:-torch}"
-DATA_CACHE_PATH="${DATA_CACHE_PATH:-/home/cache}"
```
Have you accidentally removed `DATA_CACHE_PATH`?

nsakkine replied:
See below

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -231,24 +207,48 @@ DATA_ARGS="
 "
-
-# For multi-node runs DATA_CACHE_PATH should point to a common path accessible by all the nodes (for eg, an NFS directory)
 DATA_CACHE_PATH="/home/cache"
```
Shall we instead remove this hardcoded var and return back the var definition with default `"/home/cache"` value?

nsakkine replied:
Thanks for spotting this, it is a mistake. I think it would be even better if that would be a required variable for any multi-node run (like the rest of the script intends). Because it should point to a mounted (shared) directory, I do not find it meaningful to have a default value for it.

aivanni replied:
well `/home/cache` can be that directory if it is mounted to  `/home/cache` on every worker

nsakkine replied:
Yes, agreed and made `/root/cache` the default `DATA_CACHE_PATH`.

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -205,13 +206,22 @@ DATA_ARGS="
     echo "Using Mock data"
 else
-    DATA_ARGS="$DATA_ARGS --data-path $DATA_PATH"
+    DATA_ARGS="$DATA_ARGS --data-path $DATA_PATH --data-cache-path ${DATA_CACHE_PATH}"
```
Also these changes to `if [ -z ${DATA_PATH+x} ]; then` block were only added for llama2 but not llama3 script

nsakkine replied:
They are redundant, fixed this by removing them also in llama2 script

nsakkine committed:
Fixed DATA_CACHE_PATH bug

aivanni commented on examples/llama/train_llama2.sh:
```diff
@@ -375,4 +368,10 @@ rm tmp.txt
\ No newline at end of file
+rm tmp.txt
+
+# Extract memory usage
```
mem usage code is now duplicated

nsakkine replied:
True, fixed this

nsakkine committed:
Removed duplicate mem-usage code

aivanni commented on examples/llama/readme.md:
```diff
@@ -211,23 +184,6 @@ To run training on multiple nodes, launch the Docker container on each node. Exa
-- **BS:**
-  Global Batch size
-
-- **TP:**
```
Maybe keep `TP` section?

nsakkine replied:
Yes, done

nsakkine committed:
Restored TP docs

aivanni commented on examples/llama/readme.md:
````diff
@@ -134,50 +147,10 @@ You can use either mock data or real data for training.
-  TEE_OUTPUT=1 MBS=2 BS=256 TP=1 TE_FP8=1 SEQ_LENGTH=8192 MODEL_SIZE=8  MASTER_ADDR=IP_NODE0 NNODES=2 NODE_RANK=1 bash examples/llama/train_llama3.sh
-  ```
----
+  **Note** that for multi-node runs, set `DATA_CACHE_PATH` to a directory accessible to
````
Isn't this part more relevant to belong to the `Multi-node Training` section?

nsakkine replied:
Yes, moving this

aivanni commented on examples/llama/readme.md:
```diff
@@ -104,6 +103,20 @@ export GLOO_SOCKET_IFNAME=ens50f0np0
 
 in the training scripts based on the output.
 
+**Note** that for multi-node runs, make sure that the correct network drivers are
```
The previous variant was explicit in saying `"make sure correct network drivers are installed on the nodes"` . In this variant we say "`either or"` saying that it is an option to install drivers on the node. What is correct? We must have drivers on the nodes or having them in docker is enough?

aivanni replied:
Also shall this Note be moved to `Multi-node Training` section? I feel that division between single-node and multi-node is top level choice users make when consulting documentation and it is better to aggregate all details about multi-node training in that section (where they maybe are not Notes but suggestions rather)

nsakkine replied:
@vidushi8 could you help us here and tell which form is better

vidushi8 replied:
As mentioned in previous comments, I think it is much easier to explain all this setup at one place in the multi-node configuration section as present currently: https://github.com/ROCm/Megatron-LM/blob/rocm_dev/examples/llama/readme.md#24-multi-node-training

Lets keep the original version, were we explicitly talk about the drivers. They have to be installed correctly in the nodes.

aivanni commented on examples/llama/prepare_dataset.sh:
```diff
@@ -1,43 +1,46 @@
+    --input "${DATASET_PATH}"  \
+    --tokenizer-type $TOKENIZER_TYPE \
+    --tokenizer-model "${TOKENIZER_MODEL}" \
+    --output-prefix "${DATA_DIR}/data" \
```
Maybe we want that `--output-prefix` has `$DATASET` part in the name?

nsakkine replied:
Agreed to keep as is

aivanni commented on examples/llama/train_llama3.sh:
```diff
@@ -88,64 +93,39 @@ fi
-  mkdir -p $TOKENIZER_MODEL
-  HF_TOKEN="${HF_TOKEN:-hf_xxxx}" #set huggingface access token to be able to download tokenizer
-  wget --header="Authorization: Bearer $HF_TOKEN" -O $TOKENIZER_MODEL/special_tokens_map.json https://huggingface.co/meta-llama/Llama-3.1-8B/resolve/main/special_tokens_map.json
-  wget --header="Authorization: Bearer $HF_TOKEN" -O $TOKENIZER_MODEL/tokenizer.json https://huggingface.co/meta-llama/Llama-3.1-8B/resolve/main/tokenizer.json
```
Overall the feel is that we cleaned out all details about tokenizers that were scattered over `prepare_dataset.sh`, `readme.md` and `train_llamaX.sh `. E.g. what are different types of tokenizers `GPT2BPETokenizer`, `HuggingFaceTokenizer`, `Llama2Tokenizer`, what files each of them consists of and examples and urls how they can be downloaded. So we went from assuming zero prerequisite knowledge about tokenizers, to knowing details of tokenizers and how to get them. While the code base was polished nicely, I think we should add a section in readme to document all those pieces and examples about tokenizers to enable single place on-boarding.

vidushi8 reviewed (changes requested):
@nsakkine can you update the readme to reflect the latest multi-node setup/examples which was recently merged into rocm-dev. I highly suggest keeping separate sections for single-node and multi-node examples/setups/runs as is there in the current rocm-dev branch. 
Feel free to change the tokenizer or dataset preparation section as per your changes in the PR but please do not remove content/examples for multinode setup or FSDP enablement.

nsakkine committed:
Fixed default DATA_CACHE_PATH. Updated README accordingly and added
a section about tokenizers.

vidushi8 commented on examples/llama/readme.md:
```diff
@@ -1,140 +1,189 @@
+and similarly for `examples/llama/train_llama3.sh`. 
+
+For either script, to run training with non-default options, simply add arguments as
+shown in
```
Please explicitly state this command is to enable FSDP-v2 like it exists in current readme:
`To run the training with FSDP-v2 enabled, simply add FSDP=1 argument, for example, use the following command:`

nsakkine replied:
Done

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,140 +1,189 @@
 ```bash
-export NCCL_SOCKET_IFNAME=ens50f0np0
-export GLOO_SOCKET_IFNAME=ens50f0np0
+MBS=2 BS=16 TP=1 TE_FP8=0 FSDP=1 RECOMPUTE=1 SEQ_LENGTH=8192 ./examples/llama/train_llama3.sh
````
Please include MODEL_SIZE parameter and bash. For user environment, it may not be executable script by default
`MBS=2 BS=16 TP=1 TE_FP8=0 FSDP=1 RECOMPUTE=1 SEQ_LENGTH=8192 MODEL_SIZE=70 bash examples/llama/train_llama3.sh`

nsakkine replied:
Reverted to `rocm_dev` branch regarding these

vidushi8 commented on examples/llama/readme.md:
```diff
@@ -1,140 +1,189 @@
 
-## 2. Configurations in Script (`Megatron-LM/examples/llama`)
-Use `train_llama3.sh` for Llama3/3.1 models and `train_llama2.sh` for Llama2 models.
+## 2. How to Run
```
I think it is a better flow for user to learn about Configuration in section 2 and then place How to run in section 3

nsakkine replied:
Yes, I agree

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,140 +1,189 @@
   ```bash
-  MOCK_DATA=0
-  DATA_PATH="/data/bookcorpus_text_sentence" # Change to where your dataset is stored
+  MASTER_ADDR=address NNODES=N NODE_RANK=0 ./examples/llama/train_llama2.sh
````
Please use richer examples as in original readme : https://github.com/ROCm/Megatron-LM/blob/rocm_dev/examples/llama/readme.md#32-multi-node-training
Better to replicate the 2node example from original readme, which has been verified to work.

nsakkine replied:
Done

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,140 +1,189 @@
-  Use the `HuggingFaceTokenizer`. Set the HuggingFace model path in the `TOKENIZER_MODEL` variable:
   ```bash
-  TOKENIZER_MODEL=meta-llama/Llama-3.1-8B  # For Llama3
+  MASTER_ADDR=address NNODES=N NODE_RANK=i ./examples/llama/train_llama2.sh
````
Better to replicate the 2node example from original readme, which has been verified to work. 
In the above case, the default Global batch size will be distributed in N nodes, and it may not work with any arbitrary N nodes. It has to be divisible. That's why keeping a concrete 2 node example is better. 

We can explain the parameters NNODES, NODE_RANK, etc in detail in the multi-node configuration.

nsakkine replied:
Agreed and done

vidushi8 commented on examples/llama/readme.md:
```diff
@@ -1,140 +1,189 @@
+**Note** that for multi-node runs, remember to properly setup a bind mount, with the
+default mount point `/root/cache` inside the container, to a host directory accessible
+to all of the nodes, for example a NFS directory. For non-default mount points, set
+`DATA_CACHE_PATH` appropriately and pass it to the training scripts.
```
I think it is much easier to explain all this setup at one place in the multi-node configuration section as present currently: https://github.com/ROCm/Megatron-LM/blob/rocm_dev/examples/llama/readme.md#24-multi-node-training

nsakkine replied:
Yes, moved these details there

vidushi8 commented on examples/llama/readme.md:
````diff
@@ -1,140 +1,189 @@
-To run the training on a single node, go to Megatron-LM folder, use the following command:
 ```bash
-TEE_OUTPUT=1 MBS=2 BS=128 TP=1 TE_FP8=1 SEQ_LENGTH=8192 MODEL_SIZE=8  bash examples/llama/train_llama3.sh
+export NCCL_IB_HCA=rdma0,rdma1,rdma2,rdma3,rdma4,rdma5,rdma6,rdma7
````
This is also relevant to only multinode configuration. Please keep it inside Multi-node Training configuration section.

nsakkine replied:
Done

vidushi8 reviewed (commented):
@nsakkine added some more clarification. It looks mostly good to me once the readme is fixed.

nsakkine committed:
Revised README according to comments

nsakkine committed:
Fixed README

nsakkine committed:
Fixed README

vidushi8 reviewed (approved):
LGTM, thanks @nsakkine.

aivanni reviewed (approved):
Looks great to me.
